### PR TITLE
Refactoring, extract common functionality, fix bugs, add tests, add CI.

### PR DESCRIPTION
**Pull request by ibrt**

### CODE REVIEW

File: kit/utils/observer.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// Observer describes the observer pattern.
+type Observer interface {
+	Publish(eventName string, eventData interface{}) error
+	Subscribe(eventName string, eventSubscriber EventSubscriber)
+}
+
+// EventSubscriber is a callback function for observed events.
+type EventSubscriber func(eventName string, eventData interface{}) error
+
+var noSubscribers = []EventSubscriber{}
```
Comment (vood) [inline kit/utils/observer.go]: over engineering? I think it is better to return nil immediately from `Publish` when there are no subscribers

File: kit/utils/observer.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// Observer describes the observer pattern.
+type Observer interface {
+	Publish(eventName string, eventData interface{}) error
+	Subscribe(eventName string, eventSubscriber EventSubscriber)
+}
+
+// EventSubscriber is a callback function for observed events.
+type EventSubscriber func(eventName string, eventData interface{}) error
```
Comment (vood) [inline kit/utils/observer.go]: I don't think we need to leak eventName into the subscriber.

File: kit/utils/observer.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+func NewSyncObserver() Observer {
+	return &syncObserver{
+		mutex:       &sync.Mutex{},
+		subscribers: make(map[string][]EventSubscriber),
+	}
+}
+
+// Publish implements Observer interface.
+func (o *syncObserver) Publish(eventName string, eventData interface{}) error {
+	subscribers := o.safeGetSubscribers(eventName)
```
Comment (vood) [inline kit/utils/observer.go]: Good catch

File: kit/utils/observer.go (hunk 2 above)
Comment (ibrt) [inline kit/utils/observer.go]: It's a more flexible API, now you can have the same subscriber function for different event types.

File: kit/utils/observer.go (hunk 1 above)
Comment (ibrt) [inline kit/utils/observer.go]: So the reason for this is that I wanted to acquire the log in a separate method so I could defer unlock. But good point, I changed it to return nil and short circuit in the caller.

**Pull request merged**